snapcraft package #765

andykimpe opened the pull request:
add snap package for demand issues #765

https://github.com/TASEmulators/desmume/issues/765

rofl0r commented on snapcraft.yaml:
```diff
@@ -0,0 +1,81 @@
+name: desmume-emulator
+version: 0.9.11
```
current release is 0.9.13.

andykimpe commented:
actual latest deb version stable ubuntu 0.9.11

https://packages.ubuntu.com/search?keywords=desmume

Paquet desmume

    [focal (20.04LTS)](https://packages.ubuntu.com/focal/desmume) (games): Nintendo DS emulator [universe]
    0.9.11-3build1: amd64 arm64 armhf ppc64el riscv64 s390x
    [jammy (22.04LTS)](https://packages.ubuntu.com/jammy/desmume) (games): Nintendo DS emulator [universe]
    0.9.11-3build2: amd64 arm64 armhf ppc64el riscv64 s390x
    [lunar (23.04)](https://packages.ubuntu.com/lunar/desmume) (games): Nintendo DS emulator [universe]
    0.9.11-4.1: amd64 arm64 armhf ppc64el riscv64 s390x
    [mantic (23.10)](https://packages.ubuntu.com/mantic/desmume) (games): Nintendo DS emulator [universe]
    0.9.11-4.1: amd64 arm64 armhf ppc64el riscv64 s390x
    [noble](https://packages.ubuntu.com/noble/desmume) (games): Nintendo DS emulator [universe]
    0.9.11-7: amd64 arm64 armhf ppc64el riscv64 s390x

I'm busy looking for 0.9.13

andykimpe commented:
for Ubutun Debian and snapcraft require you solve problem

0.9.13 no build with gtk3 in autotools

https://github.com/TASEmulators/desmume/issues/768

rofl0r commented:
btw you can get current ubuntu-built gtk2 version here: https://github.com/TASEmulators/desmume/actions/runs/7655965192 scroll down for "artifacts". the master version is better anyway as a crucial bug in gtk code has been fixed.

rofl0r commented:
to see how it was built check the files in .github. -glade isn't supported atm; afaik the original glade-generated frontend has been manually improved and is now available via gtk2 (autoconf) and gtk3 (meson).
btw if there's a remote chance this will be merged, it will require you to remove the debian stuff you added. if it's just here in your fork for the purpose of getting the snap built once, it's fine though.

andykimpe commented:
ok debian file removed and updated to version 0.9.13 for snap

local build tested and work

online build in progress for update

https://snapcraft.io/desmume-emulator/builds

rofl0r commented on snapcraft.yaml:
```diff
@@ -0,0 +1,113 @@
+    source: https://github.com/TASEmulators/desmume/archive/refs/tags/release_0_9_13.tar.gz
+    plugin: nil
+    override-build: |
+      LDFLAGS="" CFLAGS="" FFLAGS="" FCFLAGS="" DEB_CFLAGS_SET="" DEB_CXXFLAGS_SET="" DEB_FFLAGS_SET="" DEB_FCFLAGS_SET="" DEB_LDFLAGS_SET="" meson setup --prefix=${CRAFT_PART_INSTALL}/usr desmume/src/frontend/posix/build desmume/src/frontend/posix
```
should probably set CFLAGS to -O2 or -O3, else desmume will be very slow

andykimpe replied:
ok I suspect that there are still a few things to correct

rofl0r commented on snapcraft.yaml:
```diff
@@ -0,0 +1,113 @@
+      - libgcc-s1      
+      - libgdk-pixbuf-2.0-0 
+      - libgl1
+      - libglade2-0 
```
glade isnt needed, if you dont build the glade frontend (which is defunct anyway)

andykimpe replied:
for the dependency packages I first used what was in the 0.9.13 package then added but I did not remove what was no longer useful

rofl0r commented on snapcraft.yaml:
```diff
@@ -0,0 +1,113 @@
+      - libglade2-0 
+      - libglib2.0-0 
+      - libglu1-mesa
+      - libgtk2.0-0 
```
you dont need to install gtk2-related packages if you use the meson recipe, it builds only gtk+3 support

rofl0r commented on snapcraft.yaml:
```diff
@@ -0,0 +1,113 @@
+      - libgtkglext1
+      - libosmesa6 
+      - libpango-1.0-0 
+      - libsdl1.2debian 
```
why sdl 1.2 ? we use sdl2.

andykimpe replied:
ok i replace by libsdl2-dev

https://packages.ubuntu.com/jammy/libsdl2-dev

rofl0r commented on snapcraft.yaml:
```diff
@@ -0,0 +1,113 @@
+      - g++
+      - gcc
+      - gettext
+      - make
```
seems to me that one of build-packages vs stage-packages are packages needed for the build vs packages that need to be shipped together with the app.
you only need cmake, make, nasm, gcc, etc in the former category, not in both.

andykimpe replied:
exact and I also need to see how to replace the dev with their standard package

andykimpe replied:
and I believe that the patches from the rpmfusion team of Fedora are precisely for the architectures which were released in failure of construction so I will see to add them

https://github.com/rpmfusion/desmume/blob/master/desmume-0.9.13-aarch64.patch

https://github.com/rpmfusion/desmume/blob/master/desmume-0.9.13-arm.patch

https://github.com/rpmfusion/desmume/blob/master/desmume-0.9.13-formatstring.patch

https://github.com/rpmfusion/desmume/blob/master/desmume-0.9.13-ppc64le.patch

https://github.com/rpmfusion/desmume/blob/master/desmume-0.9.13-tinyxml.patch

andykimpe commented:
ok I think it's good like that I moved it to the snap folder to add the patches it's cleaner snapcraft looks for the snapcraft.yaml file either at the root of the project or in the snap folder

I restarted the build and got what it gives
I also corrected the desktop file because the icon is not displayed

andykimpe commented:
well apparently move to the snap folder it doesn't work however that's who says it in the doc weird ok well I'm going to put it back to the root

directly it came out in error and in addition he put an lxc template of 20.04 so I noted core 22 or 22.04

andykimpe commented:
not all is going well sort failed also

However, I didn't modify much, nothing that requires that suddenly everything comes out in error

andykimpe commented:
ok

Stage package not found in part 'desmume-emulator': libgtk-3-0-udeb.

lol ok online for focal 20.04 not found on jammy 22.04

I hadn't seen it on the Ubuntu package page

https://packages.ubuntu.com/search?suite=default&section=all&arch=any&keywords=libgtk-3&searchon=names

andykimpe commented:
on certain architecture it works but others always return errors

in architecture arm64

https://snapcraft.io/desmume-emulator/builds/2378193

build error

```
:: ../../../matrix.h: At global scope:
:: ../../../matrix.h:652:13: warning: ‘void buffer_copy_or_constant_s8(void*, const void*, size_t, s8)’ defined but not used [-Wunused-function]
::   652 | static void buffer_copy_or_constant_s8(void *__restrict dst, const void *__restrict src, const size_t vecLength, const s8 c)
::       |             ^~~~~~~~~~~~~~~~~~~~~~~~~~
:: ninja: build stopped: subcommand failed.
```

in architecture s390x

https://snapcraft.io/desmume-emulator/builds/2378195

build error

```
Installing build-snaps
Failed to install or refresh snap 'gnome-42-2204'.
'gnome-42-2204' does not exist or is not available on channel 'latest/stable'.
Use `snap info gnome-42-2204` to get a list of channels the snap is available on.
Full execution log: '/root/.local/state/snapcraft/log/snapcraft-20240202-065515.791715.log'
Build failed
```

That's a problem of missing dependency on the canonical repositories, it's up to them to correct it, I'll put them back here

https://bugs.launchpad.net/snapd/+bug/2052015


same problem on ppc64el architecture

I add it to the bug

https://snapcraft.io/desmume-emulator/builds/2378194

same problem on riscv64 architecture

I add it too

https://snapcraft.io/desmume-emulator/builds/2378196

amd64 and armhf build work

https://snapcraft.io/desmume-emulator/builds/2378191

https://snapcraft.io/desmume-emulator/builds/2378192


you have to find out why it doesn't work on the arm64 architecture personally I have no idea

for other architectures in error it is up to canonical to correct the missing dependency especially since this one is not listed in the script

rofl0r commented:
> https://snapcraft.io/desmume-emulator/builds/2378193
> build error

those links can only be looked at with a login, so they're not useful for us.
if you want help getting these build errors fixed, paste full build log on a pastebin site. in the snippet you pasted there's only a harmless warning.

andykimpe commented:
I am surprised it is marked that the builds are public yet on their site


fortunately we can export it

arm64 full log

[buildlog_snap_ubuntu_jammy_arm64_43136a0fa148bffe6910546741a37632_BUILDING.txt.gz](https://github.com/TASEmulators/desmume/files/14141704/buildlog_snap_ubuntu_jammy_arm64_43136a0fa148bffe6910546741a37632_BUILDING.txt.gz)



for other architectures it's due to a missing package at your level you can't do anything about it

because it is snapcraft itself that the package and absent it is because I opened an inssue on launchpad

rofl0r commented:
24eb5ed95c6cbdaba8b3c63a99e95e899e8a5061 should fix that specific issue.

andykimpe commented:
but this patch is already here

https://github.com/andykimpe/desmume/blob/master/patches/desmume-0.9.13-aarch64.patch

or maybe snapcraft is not applying the patches correctly
